### PR TITLE
Correctly log exceptions that are thrown during cache prewarming

### PR DESCRIPTION
**Pull request by tlrx**
Today if an exception is thrown during the prewarming of (parts of) the files it is simply swallowed. This is confusing because the shard recovery state is left in a `FINALIZE` state but nothing in the logs can help troubleshooting why the recovery has been interrupted.

This pull request adds log traces to warn when an exception occurred during the prewarming of a part of a file. In the issue we 
encountered it would have logged something like this:
```
[2021-06-22T13:42:10,510][WARN ][o.e.x.s.s.SearchableSnapshotDirectory] [pop-os] [test][0] prewarming failed for file [_a.cfe]
org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution
	at org.elasticsearch.common.util.concurrent.FutureUtils.rethrowExecutionException(FutureUtils.java:80) ~[elasticsearch-7.13.3-SNAPSHOT.jar:7.13.3-SNAPSHOT]
	...
Caused by: java.util.concurrent.ExecutionException: java.io.IOException: Failed to prefetch file part in cache
	at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:262) ~[elasticsearch-7.13.3-SNAPSHOT.jar:7.13.3-SNAPSHOT]
	... 21 more
Caused by: java.io.IOException: Failed to prefetch file part in cache
	at org.elasticsearch.xpack.searchablesnapshots.store.input.CachedBlobContainerIndexInput.prefetchPart(CachedBlobContainerIndexInput.java:216) ~[?:?]
	... 3 more
Caused by: java.nio.file.NoSuchFileException: /tmp/cluster/repo/indices/vLuVOFjKRAexJYJ7h4dVXw/0/__iWOCoAtpR2-WDCZ06eOUew
	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[?:?]
	...

```

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java [hunk 1]
```diff
@@ -555,6 +563,9 @@ private void prewarmCache(ActionListener<Void> listener) {
                 }
             } catch (IOException e) {
                 logger.warn(() -> new ParameterizedMessage("{} unable to prewarm file [{}]", shardId, file.physicalName()), e);
+                if (submitted == false) {
+                    completionListener.onFailure(e);
```
Comment (henningandersen) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java:567]: I think this attempts to ensure that the recovery state moves to `DONE`? I guess that is more correct than `FINALIZE`, but I think we should add a test for it.

**Review by henningandersen — APPROVED** [review]
LGTM.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java:567]: It attempts to complete the grouped listener but has not other side effect besides logging a warning message if the shard is a relocation target: https://github.com/elastic/elasticsearch/blob/a5d2251064d3457dc253503ba9ce75dd57fa6f4e/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/SearchableSnapshotIndexEventListener.java#L78-L90

I'm OK with leaving the recovery state in `FINALIZE` as this is what allowed to detect a non-fully prewarmed shard at the first place. I think that adding more logs would have helped to troubleshoot some issues when accessing repository files.

**Comment by tlrx** [comment]
Thanks @henningandersen - I've merged it like this, see https://github.com/elastic/elasticsearch/pull/74419#discussion_r657788467

**Pull request merged**